### PR TITLE
Introduce modularity

### PR DESCRIPTION
**Pull request by gberaudo**
This PR modularizes the ngeo search directive.
See https://github.com/camptocamp/ngeo/issues/2004 for the original issue and proposition.
The second commit is an improvement to the proposition.

The directive is made available to the library itself and to client applications through an Angular module.
This is on a par with the Angular guide available at https://docs.angularjs.org/guide/module.

This is a first step, the adaptation work will be incremental, depending on the needs and contributions of client applications.

Client applications interested by this approach includes Suissealpine and Geoadmin and should bring the following contributions:
- WMTS/WMS/GPX/... import module from Geoadmin;
- Swisstopo name component/directive for Suissealpine (ported from Schweizmobil);
- Swisstopo height directive for Suissealpine.

The long term goal is having the ngeo part of the library modularized.

### CODE REVIEW

File: src/modules/search/search.js [hunk 1]
```diff
@@ -1,6 +1,13 @@
-goog.provide('ngeo.searchDirective');
+/**
+ * @module ngeo search namespace
+ */
+goog.provide('ngeo.modules.search');
```
Comment (sbrunner) [inline src/modules/search/search.js]: Why not ngeo.search?

File: src/modules/search/search.js [hunk 2]
```diff
@@ -24,7 +31,7 @@ goog.require('ngeo');
  * @ngdoc directive
  * @ngname ngeoSearch
  */
-ngeo.searchDirective = function() {
+ngeo.modules.search.Directive = function() {
```
Comment (sbrunner) [inline src/modules/search/search.js]: ngeo.search.Directive ?

File: src/ngeo.js [hunk 3]
```diff
@@ -20,9 +20,13 @@ goog.require('ol.layer.VectorTile');
 /** @suppress {extraRequire} */
 goog.require('ol.Map');
 
+goog.require('ngeo.modules.search');
 
 /** @type {!angular.Module} */
-ngeo.module = angular.module('ngeo', ['gettext', 'ui.date', 'floatThead']);
+ngeo.module = angular.module('ngeo', [
+    ngeo.modules.search.module.name,
```
Comment (sbrunner) [inline src/ngeo.js]: I don't know about that ...

File: examples/search.js [hunk 4]
```diff
@@ -17,7 +16,7 @@ goog.require('ol.source.Vector');
 
 
 /** @type {!angular.Module} **/
-app.module = angular.module('app', ['ngeo']);
+app.module = angular.module('app', [ngeo.module.name]);
```
Comment (sbrunner) [inline examples/search.js]: Shouldn't depends on ngeo.serch module ?

File: src/modules/search/search.js (hunk 1 above)
Comment (gberaudo) [inline src/modules/search/search.js]: In the prospect of migrating one day to ES6 modules I kept the naming of the filesystem.

File: src/ngeo.js (hunk 3 above)
Comment (gberaudo) [inline src/ngeo.js]: This simply get the `ngeoSearch` string from the module.

File: examples/search.js (hunk 4 above)
Comment (gberaudo) [inline examples/search.js]: The example depended on the main ngeo module. The search module is already a dependency of the ngeo module so there is no need to add it here.

**Comment by gberaudo** [comment]
@sbrunner I moved the creategeojsonbloodhound file.

**Comment by gberaudo** [comment]
@sbrunner, as discussed I renamed the files, but without `ngeo` prefixes; it was making the names too ugly.
I added a README.md explaining the guidelines for this directory. Does it look good to you?

File: src/modules/README.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+Example for the search feature:
+- an `ngeoSearchModule` module in `modules/search/searchmodule.js`;
+- an `ngeoSearch` directive (in partials `ngeo-search`) in `modules/search/search.js`;
+- an `ngeoCreateGeoJSONBloodhound` service in ``modules/search/creategeojsonbloodhound.js`.
+
+The filenames are directly derived from the names registered in Angular (take away the ngeo prefix).
+
+
+# Google library
+The code do not use the Google library with exceptions for `goog.provide`, `goog.require` and `goog.asserts.*`.
```
Comment (sbrunner) [inline src/modules/README.md]: :-)

**Comment by gberaudo** [comment]
I relaxed the filename rules to make it more pleasant to use. As a consequence, it is not yet completly ES6 module friendly but I think it is better to go slowly than rush.

@sbrunner, @fredj could you please have a last look?

File: src/modules/search/creategeojsonbloodhound.js [hunk 6]
```diff
@@ -1,10 +1,10 @@
+/**
+ * @module ngeo search namespace
+ */
+goog.provide('ngeo.search.creategeojsonbloodhound');
```
Comment (fredj) [inline src/modules/search/creategeojsonbloodhound.js]: name should be ngeo.search.createGeoJSONBloodhound

File: src/modules/search/creategeojsonbloodhound.js [hunk 7]
```diff
@@ -41,13 +41,13 @@ goog.require('ol.format.GeoJSON');
...
  * @ngdoc service
- * @ngname ngeoCreateGeoJSONBloodhound
+ * @ngname creategeojsonbloodhound
  */
 ngeo.CreateGeoJSONBloodhound;
 
 
 /**
- * @param {string} url an URL to a search service.
+ * @param {string} url an URL to a search.creategeojsonbloodhound.
```
Comment (fredj) [inline src/modules/search/creategeojsonbloodhound.js]: bad replace

**Comment by gberaudo** [comment]
Thanks for the review, I handled all comments. Now waiting for the Travis green light.

**Pull request merged**